linera-service/src/linera-exporter/runloops/block_processor.rs

```diff
@@ -0,0 +1,461 @@
+        interval.set_missed_tick_behavior(MissedTickBehavior::Delay);
+
+        loop {
+            tokio::select! {
+
+                biased;
+
+                _ = &mut pinned => break,
+
+                _ = interval.tick() => self.storage.save().await?,
+
+                Some(next_block_notification) = self.queue_front.recv() => {
+                    let walker = Walker::new(&mut self.storage);
+                    if let Err(_err) = walker.walk(next_block_notification.clone()).await {
+                        // return the block to the back of the task queue to process again later
+                        let _ = self.queue_rear.send(next_block_notification);
+                    }
+                },
+
+            }
+        }
+
+        Ok(())
+    }
+}
+
+impl<'a, S> Walker<'a, S>
+where
+    S: Storage + Clone + Send + Sync + 'static,
+{
+    fn new(storage: &'a mut BlockProcessorStorage<S>) -> Self {
+        Self {
+            path: Vec::new(),
+            visited: HashSet::new(),
+            storage,
+        }
+    }
+
+    async fn walk(mut self, block: BlockId) -> Result<(), ExporterError> {
+        if self.is_block_indexed(&block).await? {
+            return Ok(());
+        }
+
+        let node_visitor = self.get_processed_block_node(&block).await?;
+        self.path.push(node_visitor);
+        while let Some(mut node_visitor) = self.path.pop() {
+            if self.visited.contains(&node_visitor.node.block) {
+                continue;
+            }
+
+            // resolve ancestors
+            let parent = node_visitor.get_parent();
+            if let Some(parent) = parent {
+                if !self.is_block_indexed(&parent).await? {
+                    let parent_node = self.get_processed_block_node(&parent).await?;
+                    self.path.push(node_visitor);
+                    self.path.push(parent_node);
+                    continue;
+                }
+            }
+
+            // resolve siblings
+            let sibling = node_visitor.node.siblings.get(node_visitor.sibling_branch);
+            if let Some(sibling) = sibling {
+                node_visitor.sibling_branch += 1;
+                if !self.is_block_indexed(sibling).await? {
+                    let sibling_node = self.get_processed_block_node(sibling).await?;
+                    self.path.push(node_visitor);
+                    self.path.push(sibling_node);
+                    continue;
+                }
+            }
+
+            let block_id = node_visitor.node.block.clone();
+            self.visited.insert(block_id.clone());
+            self.index_block(&block_id).await?;
+        }
+
+        Ok(())
+    }
+
+    async fn get_processed_block_node(
+        &self,
+        block_id: &BlockId,
+    ) -> Result<NodeVisitor, ExporterError> {
+        let block = self.get_block(block_id.hash).await?;
+        let processed_block = ProcessedBlock::process_block(block_id.clone(), block.block());
```

Comment:
Is `ProcessedBlock` a good name? We haven't _yet_ processed that block. We can leave the name but add a comment on the struct explaining what it's for - where it's used and how